Automattic/woocommerce-payments
# RPP: Factor flags

Fixes #6679. Fixes #6680.

## Related PRs

- https://github.com/Automattic/woocommerce-payments-server/pull/3828
- https://github.com/Automattic/woocommerce-payments-dev-tools/pull/98

## Changes proposed in this Pull Request

This PR allows the client to use the factor flags, provided by the server (see the related PR) to conditionally route customers into the new payment process.

#### New request class

There's a simple `Get_Payment_Process_Factors` request class, which loads the `payment-process-config/factors` route. It's not a top-level route because the payment process might require further configuration in the future.

#### Factors definition (https://github.com/Automattic/woocommerce-payments/issues/6679)

The new `WCPay\Internal\Payment\Factor` class contains constants with all possible payment factors, as well as their descriptions. Compared to the P2, there's an additional `NEW_PAYMENT_PROCESS` factor, which is used to indicate whether even the basic payment process can be entered at all, even if all other detected factors are `false`.

#### Router

The new `WCPay\Internal\Payment\Router` class:

- Is used to load and cache factors (`get_cached_factors()`) through the `Database_Cache` class.
- Allow factors to be modified by dev tools through the `wcpay_new_payment_process_enabled_factors` filter.
- Contains the `should_use_new_payment_process()` method, which compares the provided factors by the gateway to the ones in configuration. If any of the current (gateway-provided) factors is not present in the list of allowed factors, it will return `false`.

Ideally, all present factors would also be checked within this class; however, some checks require functionality that's only present within the gateway class.

#### Gateway changes

Things got slightly changed within the base gateway class. Instead of checking for the `WCPAY_NEW_PROCESS` constant within `process_payment()`, all routing and checks are now extracted into the `new_process_payment` method, which would either return null (proceeding with the legacy process), or return the payment response.

All factors are also loaded within `new_process_payment`.  Some of the checks there do not look too beautiful, but the `Payment_Information` instance is not available that early in `process_payment`, and cannot be passed to the method.

An important note here is that routing (factors) and the actual new payment process are intentionally separate. The `process_payment` method of the payment service should not be aware of any of that.

### Unrelated changes

The PR contains some changes that are not immediately related to its main purpose:

1. There is a new `AJAX_ALLOWED_KEYS` constant in the `Database_Cache` class, which allows the cache to be refreshed during AJAX calls for specific keys. That's needed because the payment process generally happens in AJAX calls.
2. There's a small change to the DI container tests. There were some leaks from `ContainerTest` that are now prevented.

## Testing instructions

1. Check out this PR. Tinker with the payment process, and make sure that because of the current version, the new payment process is never entered.
2. Use dev tools to enable certain (or all) factors for the new payment process.
3. Try checking out, and seeing the expected new payment process exception.
4. Repeat 2 and 3 with random combinations of factors.

-------------------

- [x] Covered with tests (100% coverage 😎 )
